### PR TITLE
Changes to facilitate CFA-netCDF 

### PR DESCRIPTION
**Pull request by davidhassell**
No functional changes to cfdm, but some changes that will allow the implementation of CFA in cf-python.

* Add `units` and `calendar` attributes to `NetCDFAarray`
* Populate these  `units` and `calendar` attributes in `NetCFDRead._create_netcdfarray`
* Add `get_units` and `get_calendar` methods to all `Array` subclasses
* Add `_set_units` method to `NetCDFAarray` that sets the units and calendar during a `__getitem__` call, if they've not already been set. It does this because it can cheaply grab them from the file whilst it is open. This is needed because CFA fragment array need to know their units, in case they are different to the aggregation variable's units.
* Add `get_address` method to  `NetCDFAarray`. This returns the file address as required for a CF fragment. 'varid' is is returned if   set 'ncvar' is not set. This goes beyond the CFA conventions at present, but is there for a legacy use case of JMG, either  won't   be required for CFA>=0.6, or else will be included as an enhancement to CFA.

### CODE REVIEW

**Review by sadielbartholomew — APPROVED** [review]
I've raised one minor thought to ponder, but overall this is watertight and 'does what it says on the tin', so please merge once you are ready.

As a side note I think it would be good to not introduce further commented out code to the codebase - even if it might be relevant later it always lives on in the version control system (for better or for worse) of course! So if you agree perhaps delete the commented out blocks, but it's not a blocker.

File: cfdm/data/netcdfarray.py [hunk 1]
```diff
@@ -88,7 +90,22 @@ def __init__(
                 ``valid_max``, ``valid_range``, ``_FillValue`` and
                 ``missing_value``.
 
-                .. versionadded:: (cfdm) 1.8.2
+                .. versionadded:: (cfdm) 1.8.
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/data/netcdfarray.py]: > I am guessing the 2 got chopped off during development?

Indeed it it did, thanks.

File: cfdm/cfdmimplementation.py [hunk 2]
```diff
@@ -2321,6 +2170,10 @@ def initialise_NetCDFArray(
 
             mask: `bool`, optional
 
+            units: `str` or `None`, optional
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cfdm/cfdmimplementation.py]: How about https://github.com/NCAS-CMS/cfdm/pull/223/commits/9e43f410df430016ae6a5b57ec7d1e7a38d0eca8

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cfdm/cfdmimplementation.py]: Perfect, thanks. Marking as resolved.

**Comment by davidhassell** [comment]
> As a side note I think it would be good to not introduce further commented out code to the codebase - even if it might be relevant later it always lives on in the version control system (for better or for worse) of course! So if you agree perhaps delete the commented out blocks, but it's not a blocker.

Very good point. I have deleted the code (https://github.com/NCAS-CMS/cfdm/pull/223/commits/931bfecdd6c154cc0c85c49171b2526b4185cbdf), which I should have done before :)

**Pull request merged**